value/value.go

```diff
@@ -266,96 +280,179 @@ func (m *Map) Less(rhs *Map) bool {
+	if v == nil {
+		return false
 	}
-	m.Items = append(m.Items, Field{Name: key, Value: value})
-	i := len(m.Items) - 1
-	m.index[key] = i
-	m.order = nil
-}
-
-// Delete removes the key from the set.
-func (m *Map) Delete(key string) {
-	items := []Field{}
-	for i := range m.Items {
-		if m.Items[i].Name != key {
-			items = append(items, m.Items[i])
+	_, ok := v.(bool)
+	return ok
+}
+
+func ValueBool(v Value) bool {
+	return v.(bool)
+}
+
+func IsNull(v Value) bool {
+	return v == nil
+}
+
+type Map interface {
+	Length() int
+	Get(string) (Value, bool)
+	Set(string, Value)
+	Delete(string)
+	Iterate(func(string, Value) bool)
+}
+
+type MapString map[string]interface{}
+
+func (m MapString) Length() int {
+	return len(m)
+}
+
+func (m MapString) Get(key string) (Value, bool) {
+	val, ok := m[key]
+	return val, ok
+}
+
+func (m MapString) Set(key string, val Value) {
+	m[key] = val
+}
+
+func (m MapString) Delete(key string) {
+	delete(m, key)
+}
+
+func (m MapString) Iterate(iter func(string, Value) bool) {
+	for key := range m {
+		if !iter(key, m[key]) {
+			return
 		}
 	}
-	m.Items = items
-	m.index = nil // Since the list has changed
-	m.order = nil
 }
 
-// StringValue returns s as a scalar string Value.
-func StringValue(s string) Value {
-	s2 := String(s)
-	return Value{StringValue: &s2}
+type MapInterface map[interface{}]interface{}
+
+func (m MapInterface) Length() int {
+	return len(m)
 }
 
-// IntValue returns i as a scalar numeric (integer) Value.
-func IntValue(i int) Value {
-	i2 := Int(i)
-	return Value{IntValue: &i2}
+func (m MapInterface) Get(key string) (Value, bool) {
+	val, ok := m[key]
+	return val, ok
 }
 
-// FloatValue returns f as a scalar numeric (float) Value.
-func FloatValue(f float64) Value {
-	f2 := Float(f)
-	return Value{FloatValue: &f2}
+func (m MapInterface) Set(key string, val Value) {
+	m[key] = val
 }
 
-// BooleanValue returns b as a scalar boolean Value.
-func BooleanValue(b bool) Value {
-	b2 := Boolean(b)
-	return Value{BooleanValue: &b2}
+func (m MapInterface) Delete(key string) {
+	delete(m, key)
 }
 
-// String returns a human-readable representation of the value.
-func (v Value) String() string {
-	switch {
-	case v.FloatValue != nil:
-		return fmt.Sprintf("%v", *v.FloatValue)
-	case v.IntValue != nil:
-		return fmt.Sprintf("%v", *v.IntValue)
-	case v.StringValue != nil:
-		return fmt.Sprintf("%q", *v.StringValue)
-	case v.BooleanValue != nil:
-		return fmt.Sprintf("%v", *v.BooleanValue)
-	case v.ListValue != nil:
-		strs := []string{}
-		for _, item := range v.ListValue.Items {
-			strs = append(strs, item.String())
+func (m MapInterface) Iterate(iter func(string, Value) bool) {
+	for key := range m {
+		vk, ok := key.(string)
+		if !ok {
+			continue
 		}
-		return "[" + strings.Join(strs, ",") + "]"
-	case v.MapValue != nil:
-		strs := []string{}
-		for _, i := range v.MapValue.Items {
-			strs = append(strs, fmt.Sprintf("%v=%v", i.Name, i.Value))
+		if !iter(vk, m[key]) {
```

Comment:
why not `for key, val := range m {` to avoid the lookup?

Reply:
Good point